### PR TITLE
 Configure collation/charset for DB connection

### LINKED ISSUE #46: Connect to database with collation/character other than UTF8
In some cases other collations are needed in order to accept [Emojis and/or non-latin](https://stackoverflow.com/questions/35125933/mysql-utf8mb4-errors-when-saving-emojis) characters.

It seems that as of today, this project do not provide a way to connect to MySQL database with any other option than UTF8, which leads to database errors when trying to add those "special" new characters.

With the wide use of smartphones, its basically impossible to build any application that do not support emojis. Its used anywhere, and its popularizing even more each year, so I believe the project should support some way to handle this.

### PR DESCRIPTION
**Pull request by Dakad**
Resolves #46 

## Context

Allow the configuration of the collation to use during the DB.open connection.
Just include into the connection string URI `&encoding=utf8mb4_unicode_520_ci`
The charset is deduced from the encoding.
Only the collation with their ID below than 1 byte can be used and some collation are not supported by MySQL.
See: https://dev.mysql.com/doc/refman/5.7/en/charset-connection.html#charset-connection-impermissible-client-charset

### CODE REVIEW

**Comment by bcardiff** [comment]
If we use a different collation, should the strings sent as arguments be encoded differently? We would need a couple of specs for this change, I fear it could break the support of other collations if there are none.

**Comment by straight-shoota** [comment]
@Dakad Would it be possible to add some specs for this?

Also note the question in https://github.com/crystal-lang/crystal-mysql/pull/100#issuecomment-1023328125

**Comment by Dakad** [comment]
@straight-shoota @bcardiff  My bad, I completely forgot about this :disappointed: I will add some specs. 

About the question, I'm not sure to fully understand it. I expect the `?encoding=` param to be in UTF8 since it's a [URI::Params](https://crystal-lang.org/api/1.5.1/URI/Params.html)

File: src/mysql/collations.cr [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+    "utf8mb4_hungarian_ci":     242,
+    "utf8mb4_sinhala_ci":       243,
+    "utf8mb4_german2_ci":       244,
+    "utf8mb4_croatian_ci":      245,
+    "utf8mb4_unicode_520_ci":   246,
+    "utf8mb4_vietnamese_ci":    247,
+  }
+
+  def self.default_collation
+    "utf8_general_ci"
```
Comment (Dakad) [inline src/mysql/collations.cr:156]: I reverted the default collation to be `utf8_general_ci` as before.

File: spec/driver_spec.cr [hunk 2]
```diff
@@ -30,13 +30,28 @@ describe Driver do
     DB.open "mysql://crystal_test:secret@#{database_host}/crystal_mysql_test" do |db|
       db.scalar("SELECT DATABASE()").should eq("crystal_mysql_test")
       db.scalar("SELECT CURRENT_USER()").should match(/^crystal_test@/)
+      db.scalar("SELECT @@collation_connection").should eq("utf8_general_ci")
```
Comment (Dakad) [inline spec/driver_spec.cr:33]: I added these 2 assertions to check when the encoding is not provided in the connection string, the DB connection's collation should be utf8_general_ci.

Not sure if it's a good practice to have that. I assume the DB server could have a different collation set in the config.ini

Comment (Dakad) [reply spec/driver_spec.cr]: I added another section to the README.md

**Review by straight-shoota — APPROVED** [review]

File: README.md [hunk 3]
```diff
@@ -77,3 +77,17 @@ Then use the example above changing the `DB.open` line to
 ```crystal
 DB.open "mysql://test:yourpassword@localhost/test" do |db|
 ```
+
+### Connection URI
+
+The connection string has the following syntax:
+
+```
+mysql://[user[:[password]]@]host[:port][/schema][?param1=value1&param2=value2
```
Comment (beta-ziliani) [inline README.md:86]:
```suggestion
mysql://[user[:[password]]@]host[:port][/schema][?param1=value1&param2=value2]
```

**Review by beta-ziliani — COMMENTED** [review]
Minor fix. I'll commit it directly to not bother the contributor for such a minion task

**Review by beta-ziliani — APPROVED** [review]

**Review by bcardiff — APPROVED** [review]
I don't think we need a default_charset function. Seems it's not used.

The specs do not verify if a string column is able to store and retrieve a string with special chars. If this is already working we can add those specs later and add the feature. But it does not seems we are preventing to break this in the future as is.

**Comment by bcardiff** [comment]
I checked with other native driver implementation and it seems they don't do anything to the payload if the collation/charset is different. So extra 👍 on that regard.

**Comment by straight-shoota** [comment]
All right, then let's merge it 👍

**Comment by bcardiff** [comment]
Yup. I would drop the unneeded defs I pointed above and merge it. I can take care of it in the following days.

**Comment by bcardiff** [comment]
Thanks @Dakad for the contribution and my apologies for the long delay in this PR

**Pull request merged**